File: rd-framework/src/test/kotlin/com/jetbrains/rider/framework/test/util/TestWire.kt
package com.jetbrains.rider.framework.test.util

import com.jetbrains.rider.framework.RdId
import com.jetbrains.rider.framework.AbstractBuffer
import com.jetbrains.rider.framework.base.WireBase
import com.jetbrains.rider.framework.createAbstractBuffer
import com.jetbrains.rider.util.Queue
import com.jetbrains.rider.util.reactive.IScheduler

class TestWire(scheduler : IScheduler) : WireBase(scheduler) {
    lateinit var counterpart : TestWire

    val msgQ = Queue<RdMessage>()

    var bytesWritten: Long = 0

    init {
        connected.set(true)
    }


    override fun send(id: RdId, writer: (AbstractBuffer) -> Unit) {
        require(!id.isNull)

        val ostream = createAbstractBuffer()
        writer(ostream)

        bytesWritten += ostream.position

        ostream.position = 0

        msgQ.offer(RdMessage(id, ostream))
        if (autoFlush) processAllMessages()
    }


    var autoFlush : Boolean = true
        get() = field
        set(value) {
            field = value
            if (value) processAllMessages()
        }

    fun processAllMessages() {
        while (!msgQ.isEmpty()) processOneMessage()
    }

    fun processOneMessage() {
        val msg = msgQ.poll() ?: return
        counterpart.messageBroker.dispatch(msg.id, msg.istream)
    }
}


class RdMessage (val id : RdId, val istream : AbstractBuffer)

File: rd-cpp/rd_framework_cpp/src/test/util/TestWire.cpp
//
// Created by jetbrains on 25.07.2018.
//

#include "TestWire.h"

TestWire::TestWire(IScheduler const *const scheduler) : WireBase(scheduler) {
    this->connected.set(true);
}

void TestWire::send(RdId const &id, std::function<void(Buffer const &buffer)> writer) const {
    assert(!id.isNull());
    Buffer ostream(10);
    writer(ostream);

    bytesWritten += ostream.get_position();

    ostream.rewind();

    msgQ.push(RdMessage(id, std::move(ostream)));
    if (auto_flush) {
        process_all_messages();
    }
}

void TestWire::process_all_messages() const {
    while (!msgQ.empty()) {
        process_one_message();
    }
}

void TestWire::process_one_message() const {
    if (msgQ.empty()) {
        return;
    }
    auto msg = std::move(msgQ.front());
    msgQ.pop();
    counterpart->message_broker.dispatch(msg.id, std::move(msg.buffer));
}

void TestWire::set_auto_flush(bool value) {
    auto_flush = value;
    if (value) {
        process_all_messages();
    }
}
